### PR TITLE
Config validation

### LINKED ISSUE #21: Throw Exception on invalid config
I've noticed if you have an error in your config, we don't `die`, and if that goes unnoticed, it's very confusing! I think we should a big bad red error for this!

-------------

Acceptance criteria:

- [ ] Change `E_USER_WARNING` to `E_USER_ERROR` in `get_json_file_contents_as_array()` on line 123
- [ ] Add text to the error message before `json_last_error_msg()` saying "`composer.json` could not be parsed\\n"
- [ ] Validate module names under the modules property against registered modules and allow unknown modules if they have an `entrypoint` property, trigger an error if invalid module found
- [ ] Check module names against top level config properties eg. `extra.altis.<module>` and trigger an error if a module name is found with instructions to move it under `extra.altis.modules.<module>`
- [ ] On local environments instead of just triggering an error throw an exception with the same error message

https://github.com/humanmade/altis-core/blob/f0d4e1a1230a149d6c84c99756657cbf7bdcdd71/inc/namespace.php#L119-L124

Use `Altis\Module::get_all()` to get all registered modules.

Examples of invalid configs:

Unregistered module with no "entrypoint" property:
```json
{
  "extra": {
    "altis": {
      "modules": {
        "enhanced-search": { "enabled": false }
      }
    }
  }
}
```

Registered module config not under "modules" property:
```json
{
  "extra": {
    "altis": {
      "search": { "enabled": false }
    }
  }
}
```

Custom module at top level with "entrypoint" property:
```json
{
  "extra": {
    "altis": {
      "custom-module": { "entrypoint": "path/to/file.php" }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by pdewouters**
fixes #21 

The errors get logged to the PHP error log, but may still cause a white screen because of the error level

### CODE REVIEW

**Comment by roborourke** [comment]
If they're not logged I'm not sure we can say it's solved - I'm not sure I follow what you mean in your comment to be honest.

There was already an open PR for this here #81 - I'll check over what you have here but we may want to take elements of what you've done and apply it there to complete the work.

File: inc/namespace.php [hunk 1]
```diff
@@ -135,13 +135,48 @@ function get_json_file_contents_as_array( $file ) : array {
 
 	if ( json_last_error() !== JSON_ERROR_NONE ) {
 		// phpcs:ignore
-		trigger_error( json_last_error_msg(), E_USER_WARNING );
+		trigger_error( 'composer.json could not be parsed\n' . json_last_error_msg(), E_USER_WARNING );
```
Comment (roborourke) [inline inc/namespace.php:138]: Worth noting that the newline character `\n` will not be parsed as a newline unless it's inside double quotes

File: inc/namespace.php [hunk 2]
```diff
@@ -135,13 +135,48 @@ function get_json_file_contents_as_array( $file ) : array {
...
 	return $contents;
 }
 
+/**
+ * Throws exception or triggers error depending on environment.
+ *
+ * @param string $msg
+ * @param int $level
+ */
+function log_error( string $msg, int $level = E_USER_ERROR ) {
```
Comment (roborourke) [inline inc/namespace.php:172]: `E_USER_ERROR` will cause PHP to stop executing, hence why we use `E_USER_WARNING` generally

**Comment by pdewouters** [comment]
@roborourke sorry, I didn't see that existing PR. I don' think the issue was in the correct column

The errors are actually logged and visible via `composer server logs php` 
`php_1              | [06-Jul-2020 14:52:41 UTC] PHP Fatal error:  Uncaught RuntimeException: Please specify an entry point setting for your custom module. in /usr/src/app/packages/core/inc/namespace.php:174`

File: inc/namespace.php (hunk 1 above)
Comment (rmccue) [reply inc/namespace.php:138]: We should avoid newlines in error messages anyway (as they can make parsing harder later), suggest:
```suggestion
		trigger_error( sprintf( 'composer.json could not be parsed: %s', json_last_error_msg() ), E_USER_WARNING );
```

**Comment by jazzsequence** [comment]
Updated this PR to fix the couple things that came out of code review.

**Comment by roborourke** [comment]
Ah sorry @jazzsequence - I was looking at #81 as the solution here, it goes into a bit more depth and uses the merged config to determine validity. I should have closed this one out.

**Comment by jazzsequence** [comment]
@roborourke no worries, I saw it sitting in the code review column so I thought I'd take a look. I'll peek over there. 👀

**Comment by roborourke** [comment]
Closing in favour of #81

**Pull request closed without merging**